pkg/library/flatten/network/fetch.go

```diff
@@ -0,0 +1,55 @@
+//
+// Copyright (c) 2019-2021 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package network
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	dw "github.com/devfile/api/pkg/apis/workspaces/v1alpha2"
+	"sigs.k8s.io/yaml"
+)
+
+type HTTPGetter interface {
+	Get(location string) (*http.Response, error)
+}
+
+func FetchDevWorkspaceTemplate(location string, httpClient HTTPGetter) (*dw.DevWorkspaceTemplateSpec, map[string]string, error) {
+	resp, err := httpClient.Get(location)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to fetch file from %s: %w", location, err)
+	}
+	defer resp.Body.Close() // ignoring error because what would we even do?
+	if resp.StatusCode != http.StatusOK {
+		return nil, nil, fmt.Errorf("could not fetch file from %s: got status %d", location, resp.StatusCode)
+	}
+	bytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not read data from %s: %w", location, err)
+	}
+
+	// Assume we're getting a devfile, not a DevWorkspaceTemplate (TODO: Detect type and handle both?)
+	devfile := &Devfile{}
+	err = yaml.Unmarshal(bytes, devfile)
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not unmarshal devfile from response: %w", err)
+	}
+
+	dwt, err := ConvertDevfileToDevWorkspaceTemplate(devfile)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to convert devfile to DevWorkspaceTemplate: %s", err)
+	}
+
+	return &dwt.Spec, dwt.Labels, nil
```

Comment:
Nobody has decided how to represent metadata in the transition from devfile -> devworkspace, since devfiles can have top-level attributes but devworkspaces do not. As a workaround, I'm using labels to store things like publisher, version, etc., even though labels do not match up with attributes (since attributes can be arbitrary yaml)

Reply:
Though to be fair, this data is currently unused and a holdover from compatibility checking

Reply:
Though we can't reliably deserialize _anything_ correctly since we're deserializing to the devfile/api as of whatever commit is in the Makefile, and v1alpha2 is not linked to schemaVersion :shrug:

Reply:
got that it's still open question. I'm OK with proceeding as is, it's not really related to the current PR.